internal/yggdrasil/yggdrasil.go

```diff
@@ -94,26 +96,45 @@ func NewYggdrasilHandler(deviceRepository edgedevice.Repository, deploymentRepos
 	}
 }
 
-func isRegistrationURL(url *url.URL) bool {
-	parts := strings.Split(url.Path, "/")
-	if len(parts) == 0 {
+func IsOwnDevice(ctx context.Context, deviceID string) bool {
+	if deviceID == "" {
 		return false
 	}
 
-	last := parts[len(parts)-1]
-	return last == "registration"
+	val, ok := ctx.Value(AuthzKey).(string)
+	if !ok {
+		return false
+	}
+	return val == strings.ToLower(deviceID)
 }
 
-func (h *Handler) GetAuthType(r *http.Request) int {
+// GetAuthType returns the kind of the authz that need to happen on the API call, the options are:
+// YggdrasilCompleteAuth: need to be a valid client certificate and not expired.
+// YggdrasilRegisterAuth: it is only valid for registering action.
+func (h *Handler) GetAuthType(r *http.Request, api *APIoperations.FlottaManagementAPI) int {
 	res := YggdrasilCompleteAuth
-	if isRegistrationURL(r.URL) {
-		res = YggdrasilRegisterAuth
+	if api == nil {
+		return res
+	}
+
+	route, _, matches := api.Context().RouteInfo(r)
+	if !matches {
+		return res
+	}
+
+	if route != nil && route.Operation != nil {
+		if route.Operation.ID == YggrasilAPIRegistrationOperation {
+			return YggdrasilRegisterAuth
+		}
 	}
 	return res
 }
 
 func (h *Handler) GetControlMessageForDevice(ctx context.Context, params yggdrasil.GetControlMessageForDeviceParams) middleware.Responder {
 	deviceID := params.DeviceID
+	if !IsOwnDevice(ctx, deviceID) {
```

Comment:
I'm debating with myself if an empty device-ID should be bad request or device-forbidden.
Can the device-id even be empty if it is part of the URL?

Reply:
deviceID cannot be empty, is one validation on swagger already :-)